### ISSUE TITLE
Nodes and entities have dynamic shadows even if there is no sun

### ISSUE THREAD

**Issue by erlehmann** [minetest/minetest#11972]
##### Minetest version
Commit 80d12dbe

##### OS / Hardware
Operating system: Debian GNU/Linux 11.2
CPU: Intel Core i7 8565U @ 1.80 Mhz

GPU model: Intel WhiskeyLake-U GT2 [UHD Graphics 620] (rev 02)
OpenGL version: 4.6 (Compatibility Profile) Mesa 20.3.5

##### Summary
The Mineclonia map generator generates a place called the End dimension.

It is intended to be a timeless place where the sky is the color of TV static.

As the Minetest engine does not support dimensions, the End is located deep underground.

The sun is explicitly disabled by `player:set_sun({visible = false , sunrise_visible = false})`.

Nevertheless, nodes and entities have shadows, even deep underground and with the sun explicitly disabled.

It is likely that this also affects also MineClone2 and MineClone5, as those games also have the same End dimension.

##### Steps to reproduce
1. In Mineclonia with dynamic shadows enabled, teleport to the end with `/teleport 0,-27000,0`.
2. In case you spawn in solid ground, use flying and noclip to get out of it.
3. Set the time to 15:00 using the command `/time 15:00`.
4. Notice that nodes and entities cast shadows despite no sun being visible.

##### Bugfix
Dynamic shadows should not be cast by the sun or moon if the sun or moon are not visible for a player.

**Comment by x2048** [comment]
@erlehmann Would you think #11944 will address this? With the API the shadows would not be there by default, and you would enable/set intensity in the overworld where it makes sense.

**Comment by erlehmann** [comment]
@x2048 I doubt that a possible shadow intensity API is a good solution for this issue.

Rationale:

1. Games and mods would have to add a non-trivial amount of code for any workaround here.
2. Bookkeeping is difficult and better be done by the engine to keep sun/moon and shadows in sync.
3. Seriously it makes absolutely no sense to me for an invisible sun to cause a shadow deep underground.

Mineclonia has moon phases and setting shadow intensity per player at night could be a welcome addition. But this issue is about games with several dimensions, i.e. developers can not just use the “it is always 12:00” hack that NodeCore uses IIRC.

In short, I think the shadow code is obviously buggy here and should be fixed.

**Comment by lhofhansl** [comment]
Interesting. Minetest does detect sunlight here, right? I.e. the smooth lighting would show shadows (not dynamic shadows) of things, right?
So here we "paste" over the problem that the "top" of the cave is not loaded by disabling the sun, right?
If that is the case we can indeed not expect that Minetest will this automatically, hence the game needs to disable the sun and shadows.

In fact that is why I liked controlling shadows via the sky API. Then we can do things like disabling shadows automatically when the sun is not visible.

@erlehmann I didn't follow why #11944 would not work. The game would - in addition to hiding the sun - set the shadow strength to 0, which would disable them.

**Comment by x2048** [comment]
I think, connecting shadows to presence or absence of sky bodies would be constraining. Imagine a surreal world with a fixed bright sky, no explicit sun and fixed shadows at 10am.

**Comment by lhofhansl** [comment]
@x2048 Fair enough.

**Comment by erlehmann** [comment]
> I think, connecting shadows to presence or absence of sky bodies would be constraining.

Shadows are implicitly connected to the theoretical position of the sun right now. Even if there is no sun.

Unless that is going to change, every game where the sun does not exist is going to look weird.

Those games also experiences gameplay changes for players who have shadows turned on. 

For example, there should be no way to tell time in the End dimension.

> Imagine a surreal world with a fixed bright sky, no explicit sun and fixed shadows at 10am.

It would be so much better if the initial release of shadows was just a player:set_shadows() function.

In my opinion, there should never be fixed shadows “at 10 AM” anyway. What does that even mean?

Mods should be able to hint shadows source/direction and intensity freely per player.

Not casting shadows from an invisible sun prevents bugs in several games.

Your hypothetical surreal world would be served by the shadow API.

**Comment by erlehmann** [comment]
> To my understanding, not if we assume that 'no sun == no shadows'.

IMO regardless of the resolution of this issue, a shadow API must make it possible to add a direction for each light source that should cast shadows, e.g. as a vector. Or did you not have that in mind?

**Comment by erlehmann** [comment]
> My point is that "no sun = no shadows" is a constraint, which exists in the real world, but may not necessarily apply to game worlds

I did not ask anyone to disable shadows entirely when there is no sun. I just do not want an inexistent sun to cast any shadow. If a game has no sun, but a moon, then that should be able to cast a shadow. If someone later implements players holding torches to cast shadows, that would be totally fine in the End dimension.

Also, all game worlds that have no sun *explicitly* disable the sun right now. Your hypothetical game developer who disables the sun, but still wants shadows cast a specific way would be perfectly served by a shadow API that allowed them to pass a vector for the shadow angle. That way any shadows would not be constrained by celestial objects.

> What I would like to clarify is whether we want to enforce this constraint in the engine or let the game developers decide and call player:set_lighting({shadow_intensity=0}) when necessary.

I think breaking existing games is bad. And I also think that the way you envision the shadows API – with no way to set the direction of shadows – will not give game developers the flexibility that they want.

Therefore I propose:

1. Only if sun or moon are visible, let them cast shadows. This sidesteps most issues with the End dimension or any game that is in some abstract world without a sun and/or moon.

2. MAke it possible in the shadows API to set the direction/angle of the shadows,

**Comment by erlehmann** [comment]
For the curious: The relevant logic in https://github.com/minetest/minetest/commit/0f25fa7af655b98fa401176a523f269c843d1943 can be found around these lines:

´´´
	bool is_day = timeoftheday > 0.25 && timeoftheday < 0.75;
	bool is_shadow_visible = is_day ? sky->getSunVisible() : sky->getMoonVisible();
´´´

From my understanding of the code, this means that dimensions without a sun or moon (end, nether) never have shadows.

**Comment by x2048** [comment]
Correct for now, as shadows are currently tied to the active skybody. We can implement free directional shadows at some point, and this code will go away.
